syntax = "proto3";
package cluster;

import "controller.proto";

enum MessageType {
    UNSPECIFIED = 0; // unknown command type
    ACK = 1; // acknowledged
    APPLY_TO_WORKER = 2; // apply message to worker
}

message Ack {
    string message = 1; // acknowledgment message
    string error = 2; // error message, if any
}

message AddWorker {
    string worker_id = 1; // unique worker ID
    string host = 2; // worker host
    int32 port = 3; // worker port
    int32 worker_rank = 4; // worker rank
}

message RemoveWorker {
    string worker_id = 1; // unique worker ID
}

message GetWorkersRequest {
    // No fields needed for this request
}

message GetWorkersResponse {
    repeated AddWorker workers = 1; // list of workers
}

message ApplyToWorker {
    string worker_id = 1; // unique worker ID
    controller.Message controller_message = 2; // message to apply to the worker
}

message Message {
    MessageType type = 1;
    oneof Message {
        Ack ack = 2;
        ApplyToWorker apply_to_worker = 3;
    }
}

service ClusterService {
    rpc Session(stream Message) returns (stream Message) {}
    rpc AddWorkerCommand(AddWorker) returns (Ack) {}
    rpc RemoveWorkerCommand(RemoveWorker) returns (Ack) {}
    rpc GetWorkers(GetWorkersRequest) returns (GetWorkersResponse) {}
}